File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/between/BetweenClient.kt
package org.wyrdsekai.app.engine.between

/**
 * Abstract transport for the Between network.
 * Implementations connect to NATS (via WebSocket) or other transports.
 *
 * Subject naming: "between.{zone}.{src}.{dst}.{layer}.{topic}"
 * For soul headlines: "between.household.{nodeId}.*.soul.headlines"
 *
 */
interface BetweenClient {
    val isConnected: Boolean

    /** Connect to the Between network. */
    suspend fun connect(url: String)

    /**
     * Provide NATS credentials for subsequent connects (e.g. the
     * relay_phone user from a wyrdphone:// invite).
     * Default no-op for transports without authentication.
     */
    fun setCredentials(user: String?, password: String?) {}

    /** Disconnect from the Between network. */
    suspend fun disconnect()

    /** Publish a message to a subject. */
    fun publish(subject: String, data: ByteArray)

    /** Subscribe to a subject. Returns unsubscribe function. */
    fun subscribe(subject: String, handler: (subject: String, data: ByteArray) -> Unit): () -> Unit
}

/**
 * In-memory Between client for testing.
 * Messages published are delivered synchronously to local subscribers.
 */
class InMemoryBetweenClient : BetweenClient {
    private var _connected = false
    override val isConnected: Boolean get() = _connected

    private val subscriptions = mutableListOf<Pair<String, (String, ByteArray) -> Unit>>()

    /** All messages published (for test assertions). */
    val published = mutableListOf<Pair<String, ByteArray>>()

    override suspend fun connect(url: String) {
        _connected = true
    }

    override suspend fun disconnect() {
        _connected = false
    }

    override fun publish(subject: String, data: ByteArray) {
        published.add(subject to data)
        for ((pattern, handler) in subscriptions) {
            if (subjectMatches(pattern, subject)) {
                handler(subject, data)
            }
        }
    }

    override fun subscribe(subject: String, handler: (String, ByteArray) -> Unit): () -> Unit {
        val entry = subject to handler
        subscriptions.add(entry)
        return { subscriptions.remove(entry) }
    }

    /** Simple wildcard matching: * matches one token, > matches remaining. */
    private fun subjectMatches(pattern: String, subject: String): Boolean {
        val pParts = pattern.split(".")
        val sParts = subject.split(".")
        for (i in pParts.indices) {
            if (pParts[i] == ">") return true
            if (i >= sParts.size) return false
            if (pParts[i] != "*" && pParts[i] != sParts[i]) return false
        }
        return pParts.size == sParts.size
    }
}

File: clients/rn/src/engine/between/BetweenClient.ts
/**
 * Abstract transport for the Between network.
 * Implementations connect to NATS (via NatsClient) or other transports.
 *
 */

export type BetweenMessageHandler = (subject: string, data: Uint8Array) => void;

export interface BetweenClient {
  readonly isConnected: boolean;
  connect(url: string): Promise<void>;
  disconnect(): Promise<void>;
  publish(subject: string, data: Uint8Array): void;
  subscribe(subject: string, handler: BetweenMessageHandler): () => void;
}

/**
 * In-memory Between client for testing.
 * Messages published are delivered synchronously to local subscribers.
 */
export class InMemoryBetweenClient implements BetweenClient {
  private _connected = false;
  private subscriptions: Array<{ subject: string; handler: BetweenMessageHandler }> = [];

  /** All messages published (for test assertions). */
  published: Array<{ subject: string; data: Uint8Array }> = [];

  get isConnected(): boolean {
    return this._connected;
  }

  async connect(_url: string): Promise<void> {
    this._connected = true;
  }

  async disconnect(): Promise<void> {
    this._connected = false;
  }

  publish(subject: string, data: Uint8Array): void {
    this.published.push({ subject, data });
    for (const sub of this.subscriptions) {
      if (this.subjectMatches(sub.subject, subject)) {
        sub.handler(subject, data);
      }
    }
  }

  subscribe(subject: string, handler: BetweenMessageHandler): () => void {
    const entry = { subject, handler };
    this.subscriptions.push(entry);
    return () => {
      this.subscriptions = this.subscriptions.filter(s => s !== entry);
    };
  }

  /** Simple wildcard matching: * matches one token. */
  private subjectMatches(pattern: string, subject: string): boolean {
    const pParts = pattern.split('.');
    const sParts = subject.split('.');
    for (let i = 0; i < pParts.length; i++) {
      if (pParts[i] === '>') return true;
      if (i >= sParts.length) return false;
      if (pParts[i] !== '*' && pParts[i] !== sParts[i]) return false;
    }
    return pParts.length === sParts.length;
  }
}
